### PR TITLE
Feature: Add p-less sampling

### PR DESCRIPTION
**Pull request by BII-wushuang**
## Overview

This PR contributes the p-less sampling method as documented in the paper [p-less Sampling: A Robust Hyperparameter-Free Approach for LLM Decoding](https://arxiv.org/pdf/2509.23234). This paper has been presented at the following top-tier conferences:
- ICLR 2026 Conference (Oral)
- NeurIPS 2025 Frontiers in Probabilistic Inference Workshop (Oral)

P-less sampling is a hyperparameter-free decoding strategy that calculates a dynamic threshold based on the sum of squared probabilities of the current token distribution. Tokens falling below this threshold are masked out.

## Additional Details
This PR implements the sampler across the stack:

- CPU Sampler (`src/llama-sampler.cpp`): Utilizes llama_sampler_softmax_impl to calculate probabilities, squares them to find the threshold, and masks out tokens below the threshold.
- GGML Backend (`src/llama-sampler.cpp`): Fully supports GPU execution via llama_sampler_p_less_backend_apply. It leverages ggml_soft_max, ggml_mul, ggml_sum, and ggml_step to compute the threshold and apply the mask directly on the accelerator, preventing CPU-GPU synchronization bottlenecks.
- Numerical Stability: To prevent edge-case floating-point precision errors (e.g., where a perfectly uniform distribution drops 100% of tokens due to float32 drift), a -1e-5 epsilon bias is applied to the threshold.
- Parameters (`common/common.h` / `common/arg.cpp`): Adds a boolean toggle `--p-less` to the CLI arguments, as this sampler requires no additional hyperparameters.

## Tests
We have tested the code locally for Linux and added sampling tests to `tests/test-sampling.cpp`.

Specifically, we added:
- test_p_less: Validates the exact distribution outputs.
- Edge-case arrays testing the -1e-5f epsilon safety (e.g., a uniform [0.2, 0.2, 0.2, 0.2, 0.2] distribution safely surviving truncation).
- Performance benchmarking in test_perf() to ensure the sampler meets expected speed requirements.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: Yes, AI is used for preparing sampling test examples.

### CODE REVIEW

**Comment by JaKasb** [comment]
Dear @BII-wushuang 
I tried to test the PR but failed.

1) p-less is inserted by default in the sampler chain.
Command : without enabling `--p-less`
`build/bin/llama-cli --model ../glm-47/GLM-4.7-UD-Q4_K_XL-00001-of-00005.gguf --mmap --single-turn -n 1024 -p "Write a 
haiku about Pokémon." -lv 4`
Output : 
```
slot launch_slot_: id  0 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> p-less -> ?xtc -> ?temp-ext -> dist 
slot launch_slot_: id  0 | task -1 | sampler params: 
        repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
        dry_multiplier = 0.000, dry_base = 1.750, dry_allowed_length = 2, dry_penalty_last_n = 202752
        top_k = 40, top_p = 0.950, min_p = 0.050, xtc_probability = 0.000, xtc_threshold = 0.100, typical_p = 1.000,    top_n_sigma = -1.000, temp = 1.000 mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000, adaptive_target = -1.000, adaptive_decay = 0.900
``` 

2) The state of sampler p-less is not included in the function `common_params_sampling::print()`
See : https://github.com/BII-wushuang/llama.cpp/blob/6b4e5263583eea592dc466ecef0a0c44472902d7/common/sampling.cpp#L171
See the above output.
So the user does not know whether p-less is actually enabled and working, or only inserted into the sampler chain and doing passthrough.

3) Sampler arguments needs to be documented.
See : https://github.com/ggml-org/llama.cpp/tree/master/tools/server#sampling-params

4) llama-cli and llama-server arguments do not accept parameters.
e.g. `--p-less disabled` or `--p-less 0`
Wishlist : A tunable threshold as mentioned in the paper. Is this possible with p-less-norm ?

EDIT:
5) Can you demonstrate that p-less actually influences sampling with llama.cpp ?
I failed at creating a llama-cli command that only includes the samplers : temperature and p-less.

Please feel free to reject my points in the case of user-error on my side :)

**Comment by JaKasb** [comment]
Can confirm that p-less is working. But p-less is always active, even without the flag `--p-less`.
For testing : server flags to disable default truncation samplers : `--top-k -1 --min-p -1 --top-p 2.0`

**Comment by BII-wushuang** [comment]
> Dear @BII-wushuang I tried to test the PR but failed.
> 
> 1. p-less is inserted by default in the sampler chain.
>    Command : without enabling `--p-less`
>    `build/bin/llama-cli --model ../glm-47/GLM-4.7-UD-Q4_K_XL-00001-of-00005.gguf --mmap --single-turn -n 1024 -p "Write a  haiku about Pokémon." -lv 4`
>    Output :
> 
> ```
> slot launch_slot_: id  0 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> p-less -> ?xtc -> ?temp-ext -> dist 
> slot launch_slot_: id  0 | task -1 | sampler params: 
>         repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
>         dry_multiplier = 0.000, dry_base = 1.750, dry_allowed_length = 2, dry_penalty_last_n = 202752
>         top_k = 40, top_p = 0.950, min_p = 0.050, xtc_probability = 0.000, xtc_threshold = 0.100, typical_p = 1.000,    top_n_sigma = -1.000, temp = 1.000 mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000, adaptive_target = -1.000, adaptive_decay = 0.900
> ```
> 
> 2. The state of sampler p-less is not included in the function `common_params_sampling::print()`
>    See : https://github.com/BII-wushuang/llama.cpp/blob/6b4e5263583eea592dc466ecef0a0c44472902d7/common/sampling.cpp#L171
>    See the above output.
>    So the user does not know whether p-less is actually enabled and working, or only inserted into the sampler chain and doing passthrough.
> 3. Sampler arguments needs to be documented.
>    See : https://github.com/ggml-org/llama.cpp/tree/master/tools/server#sampling-params
> 4. llama-cli and llama-server arguments do not accept parameters.
>    e.g. `--p-less disabled` or `--p-less 0`
>    Wishlist : A tunable threshold as mentioned in the paper. Is this possible with p-less-norm ?
> 
> EDIT: 5) Can you demonstrate that p-less actually influences sampling with llama.cpp ? I failed at creating a llama-cli command that only includes the samplers : temperature and p-less.
> 
> Please feel free to reject my points in the case of user-error on my side :)

Dear Jakob,
Thank you for the detailed feedback and suggestions.

I will look into all your raised points. They are all valid, and the omissions are on me as I am not familiar with the repo. Really appreciate you taking the time to highlight all these issues. Per your suggestion, I will also incorporate the Rényi k-order generalization of p-less (Eqn. 17 of the paper) but we have not delved into / derive a similar extension for p-less-norm.

**Comment by Green-Sky** [comment]
From what I can pull from the paper they did not compare against `min-p` properly. Most graphics and benchmarks seem to use the same fixed `min-p` cutoff param.

image

This simple graphic illustrates that well. The green line is relative to the first token prob, and could have been scaled to closer track the `p-less` config.

They certainly are not the same, but the paper does not really explore that much. 

image

This table seems to be the only test with differing `min-p` cutoff values. And higher test accuracy tracks with higher cutoff values, which suddenly stop at `0.2`. To get similar cutoffs to `p-less`, according to the graphic above, you would have to go to `~0.6` .
